python/ray/data/_internal/execution/streaming_executor_state.py

```diff
@@ -330,6 +339,47 @@ def select_operator_to_run(
     )
 
 
+def _try_to_scale_up_cluster(topology: Topology):
+    """Try to scale up the cluster to accomodate the provided in-progress workload.
+
+    This makes a resource request to Ray's autoscaler consisting of the current,
+    aggregate usage of all operators in the DAG + the incremental usage of all operators
+    that are ready for dispatch (i.e. that have inputs queued). If the autoscaler were
+    to grant this resource request, it would allow us to dispatch one task for every
+    ready operator.
+
+    Note that this resource request does not take the global resource limits or the
+    liveness policy into account; it only tries to make the existing resource usage +
+    one more task per ready operator feasible in the cluster.
+
+    Args:
+        topology: The execution state of the in-progress workload for which we wish to
+            request more resources.
+    """
+    # Get resource usage for all ready ops.
+    usage = ExecutionResources()
+    for op, state in topology.items():
+        usage = usage.add(op.current_resource_usage())
+        # Only include incremental resource usage for ops that are ready for
+        # dispatch.
+        if state.num_queued() > 0:
+            # TODO(Clark): Scale up more aggressively by adding incremental resource
+            # usage for more than one bundle in the queue for this op?
```

Comment:
Yeah this is probably OK for now; we can keep these optimizations for future work.